subsys/bluetooth/host/gatt.c

```diff
@@ -2011,44 +2013,46 @@ static u8_t update_ccc(const struct bt_gatt_attr *attr, void *user_data)
 
 	ccc = attr->user_data;
 
-	for (i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
-		/* Ignore configuration for different peer */
-		if (bt_conn_addr_le_cmp(conn, &ccc->cfg[i].peer)) {
-			continue;
+	/* Check if attribute requires encryption/authentication */
+	need_update = false;
+	err = bt_gatt_check_perm(conn, attr, BT_GATT_PERM_WRITE_MASK);
+	if (err) {
+		if (err == BT_ATT_ERR_WRITE_NOT_PERMITTED) {
+			BT_WARN("CCC %p not writable", attr);
+			return BT_GATT_ITER_CONTINUE;
 		}
-
-		/* Check if attribute requires encryption/authentication */
-		err = bt_gatt_check_perm(conn, attr, BT_GATT_PERM_WRITE_MASK);
-		if (err) {
-			bt_security_t sec;
-
-			if (err == BT_ATT_ERR_WRITE_NOT_PERMITTED) {
-				BT_WARN("CCC %p not writable", attr);
-				continue;
-			}
-
+		if (IS_ENABLED(CONFIG_BT_SMP)) {
 			sec = BT_SECURITY_L2;
-
 			if (err == BT_ATT_ERR_AUTHENTICATION) {
 				sec = BT_SECURITY_L3;
 			}
-
 			/* Check if current security is enough */
-			if (IS_ENABLED(CONFIG_BT_SMP) &&
-			    bt_conn_get_security(conn) < sec) {
-				if (data->sec < sec) {
-					data->sec = sec;
-				}
-				continue;
-			}
+			need_update = bt_conn_get_security(conn) < sec &&
+				      data->sec < sec;
 		}
+	}
 
-		if (ccc->cfg[i].value) {
-			gatt_ccc_changed(attr, ccc);
-			if (IS_ENABLED(CONFIG_BT_GATT_SERVICE_CHANGED) &&
-			    ccc == &sc_ccc) {
-				sc_restore(conn);
+	for (i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
+
+		/* Nothing to update */
+		if (!need_update && !ccc->cfg[i].value) {
+			continue;
+		}
+
+		/* Only update configuration for peer */
+		if (!bt_conn_addr_le_cmp(conn, &ccc->cfg[i].peer)) {
+			if (IS_ENABLED(CONFIG_BT_SMP) && need_update) {
```

Comment:
Given that I'm right in the comment above, then this part should be removed.
If I'm wrong then I think this should have had a `continue` statement here to be similar to the previous behaviour (which currently makes no sense to me).

Reply:
I understood from @fnde-ot that a peer can appear only once, so we can return as soon as we found it without changing the actual behavior.

Reply:
Right. So then it should be a `return BT_GATT_ITER_CONTINUE` here, otherwise, we will update the CCC.